Add recipe for ghc-rts.

ilitzroth opened the pull request:
### Brief summary of what the package does
Package consists of a binary module that initializes GHC
runtime system.

### Direct link to the package repository

https://github.com/ilitzroth/ghc-rts.el

### Your association with the package
Maintainer, (sole) contributor and enthusiastic user.

### Relevant communications with the upstream package maintainer
Package will compile the extension on the fly when the user calls the 
functions in this package. There might be problems with that on different
compilers/os's but that will need to be addressed when problems arise.

The  package is actually a building block for an upcoming package of mine 
which will allow one to write haskell code that can used by emacs. 
That will not require c/c++ and main use of that will be using 
development tools from emacs directly.

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

riscy commented:
Thanks for this.

Regarding the auto-compilation, I think your `(with-help-window .. (prin1-to-string` should be `(with-help-window .. (princ` -- otherwise I'm getting a blank help window.  It also seems this message will pop up every time the user updates the package, which is fine, but it might help users to change the phrasing "doesn't seem to be built on your system yet" to "hasn't been built or has been updated".

I'll also add this seems like an unorthodox use of a help window (it's not really `help` so much as a status buffer, and running e.g. `compile` might be a better choice).  I'd also recommend prompting with `yes-or-no-p` as a courtesy.

I also notice you are pre-declaring some functions some another (future?) package, e.g. `(declare-function ghc-rts::get-rts-status "ext:emacs-ghc-rts")` -- you should be aware that package-lint will complain about `::` being a non-standard separator.

Minor:
- Technically the `all` and `test` recipes should be marked phony in the Makefile, just in case
- I notice your `Created` is: `;; Created: Symbol’s function definition is void: time-stamp-string` and the `-*- lexical-binding: t -*-` trailing your Copyright notice (usually you would use the current year instead of 2025 :) could be removed

ilitzroth commented:
Thanks for your review!
* Changed the message to ".. built or updated..."
* I've made the whole thing yes-or-no-p messages
* The problem with running a compile is that is asynchronous and waiting for it 
is hairy. If you want I can do that, but I was hoping to avoid this complexity.
* the predeclared functions are the ones exported to emacs from the binary module.
If you want I can change those names to not have "::" but as now package-lint doesn't 
complain about them.
* removed second occurence of -*- lexical-binding: t -*-
* Fixed makefile and Created

riscy commented:
Apologies, I must have had a bug in my workflow last week and missed a few automatic checks --

checkdoc (using version 0.6.2):
```
Warning (emacs): 
ghc-rts.el:45: First sentence should end with punctuation
Warning (emacs): 
ghc-rts.el:45: First line should be capitalized
Warning (emacs): 
ghc-rts.el:58: Argument `force-rebuild' should appear (as FORCE-REBUILD) in the doc string
```

On first glance the docstring mentioned on line 45 looks incomplete -- the second should be an easy fix.

The last and potentially most important is to add a LICENSE file that GitHub can detect (e.g. no markup) if possible, in this case reflecting your use of the GPL -- this way people will know that the other code files (cpp, hpp, Makefile) are also GPL, but it's a good idea to (also) add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to those.

--------

With all of those out of the way, some comments on the package itself -- how far away is the GHC RTS (and more importantly the downstream Emacs packages that will use it) from completion?  I think we shouldn't be in a hurry to merge these core utilities into MELPA before there's something for them to depend upon.

ilitzroth commented:
> Apologies, I must have had a bug in my workflow last week and missed a few automatic checks --
Life is busy for everyone, no problem.
 
> checkdoc (using version 0.6.2): 
> 
> ```
> Warning (emacs): 
> ghc-rts.el:45: First sentence should end with punctuation
> Warning (emacs): 
> ghc-rts.el:45: First line should be capitalized
> Warning (emacs): 
> ghc-rts.el:58: Argument `force-rebuild' should appear (as FORCE-REBUILD) in the doc string
> ```
Fixed
> On first glance the docstring mentioned on line 45 looks incomplete -- the second should be an easy fix.
No it's just a symbol. I was going to use that to provide from the binary module but for some 
reason that doesn't work in the initialization of the binary module.

> The last and potentially most important is to add a LICENSE file that GitHub can detect (e.g. no markup) if possible, in this case reflecting your use of the GPL -- this way people will know that the other code files (cpp, hpp, Makefile) are also GPL, but it's a good idea to (also) add license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to those.

Added license & SPDX identifiers.

> With all of those out of the way, some comments on the package itself -- how far away is the GHC RTS (and more importantly the downstream Emacs packages that will use it) from completion? I think we shouldn't be in a hurry to merge these core utilities into MELPA before there's something for them to depend upon.

Well the package that will use this is basically working, but I want to distribute it with a few examples, a tutorial
and more. I can give you access to the code if you wanna have a look, it's in a private repo until it's ready
for release. It also will not be released on melpa, it should be on hackage since it will be Haskell only.

There is another package that does what I attempt to do, 
https://github.com/sergv/emacs-module
but that has some shortcomings
1) Requires C coding to start the module and as such it will always 
have a problem loading 2 modules (ghc-rts can only be intialized/deinitialized once)
unless you factor out the starting up of ghc-rts like I did here.
2) Has a difficult api which is basically undocumented. 
3) As a far as I can see it has not been picked up by anyone
4) No tutorial: https://github.com/sergv/emacs-module/issues/4
5) Does not address the hairy points of living between two garbage collectors and making
sure that memory gets freed correctly.

riscy commented:
I notice the header file has a different license than the other files (probably a typo).  Also, thanks for the pointer to the emacs-module repo and the solid explanation -- it seems there's also a haskell-emacs package that does similarly.

> it's in a private repo until it's ready for release. It also will not be released on melpa, it should be on hackage since it will be Haskell only.

That makes sense.  I think we should wait to merge this PR into MELPA until that repo is ready for release, just in case you get ambushed by other work.  Feel free to ping when that's ready!

tarsius commented:
> Feel free to ping when that's ready!

@ilitzroth ready now?

ilitzroth commented:
> > Feel free to ping when that's ready!
> 
> @ilitzroth ready now?
No, nothing on hackage yet, and probably not before the end of march since I have a 
shitload of work. The emacs package can be released as it is, since it does what it's 
supposed to do.

riscy commented:
I removed the awaiting upstream in case anyone else wants to chime in.

> I want to distribute it with a few examples, a tutorial and more. I can give you access to the code if you wanna have a look, it's in a private repo until it's ready for release.

Does it still sound fair to wait until this dependent code is ready?  Back during the initial review the package appeared brand new, but maybe things have changed.

riscy commented:
I'll mark this awaiting upstream until hearing back, but feel free to reach out.

tarsius commented:
I'm closing this now because it has been a year since this was opened. Of course that doesn't mean we cannot reopen and merge anymore.

sergv commented:
As an author of https://github.com/sergv/emacs-module and seeing the shortcomings section claiming that package is undocumented when there are clearly docs on Hackage leaves me baffled. Still at the time there was no tutorial but I managed to write one lately so that shortcoming is no longer relevant.

Speaking of

> Requires C coding to start the module and as such it will always
>have a problem loading 2 modules (ghc-rts can only be intialized/deinitialized once)
> unless you factor out the starting up of ghc-rts like I did here.

It seems to me that this problem is best addressed when there are at least 2 Haskell-based modules that someone attempts to load at the same time. I don't see why this problem needs to be adressed first unless there are modules users want to load simultaneously.

As mentioned in the shortcomings section earlier, my package doesn't appear to be picked up, which is true to my knowledge (except a module I wrote for my self but noone is interested in it), which means that there appears to be 0 Emacs modules implemented in Haskell that users would like to load. It seems redundant to have an elisp package allowing to load multiple such modules since there aren't any for a general user.
